### PR TITLE
Improve RPC handler checks and diagnostics filtering

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- reject duplicate rpc_handler registrations
- normalize diagnostics filter criteria
- test RPC multi-chunk streams and mid-stream errors
- test rpc_handler duplicate detection
- test case-insensitive diagnostics filtering

## Testing
- `make fmt`
- `make lint`
- `make test`
- `make typecheck`

------
https://chatgpt.com/codex/tasks/task_e_688a8e7a3c648322a76436dff6cbac9f

## Summary by Sourcery

Prevent duplicate RPC handler registrations, normalize diagnostics filtering criteria to be case- and path-insensitive, and add tests for RPC streaming behavior, midstream errors, duplicate detection, and diagnostics filtering.

Bug Fixes:
- Reject duplicate RPC handler registrations.

Enhancements:
- Normalize diagnostics filter criteria for case- and path-insensitive matching.

Tests:
- Add test for streaming multiple RPC results.
- Add test for handling mid-stream RPC errors.
- Add test for rejecting duplicate RPC handler registrations.
- Add test for case-insensitive diagnostics filtering.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved filtering of diagnostics to be case- and path-insensitive, ensuring consistent results regardless of input format.
  * Prevented duplicate RPC handler registration by raising an error on duplicates.

* **Tests**
  * Added tests to verify diagnostics filtering works correctly with different file name and severity casings.
  * Introduced tests for RPC handler registration to prevent duplicate handlers.
  * Expanded RPC dispatcher tests to cover streaming multiple results and error handling during streaming.
  * Refactored test code to reduce duplication and improve maintainability.

### CODE REVIEW

File: weaverd/unittests/test_rpc.py [hunk 1]
```diff
@@ -1,6 +1,7 @@
 from __future__ import annotations
 
 import builtins
+import typing as typ
```
[the remark this answers is not in the document]
Comment (leynos) [reply weaverd/unittests/test_rpc.py:4]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.

File: weaverd/unittests/test_rpc.py [hunk 2]
```diff
@@ -23,11 +36,10 @@ async def test_dispatcher_handles_registered_method() -> None:
...
 
-    request = msjson.encode({"method": "project-status"})
-    results = dispatcher.handle(request)
-    response = await builtins.anext(results)
-    assert msjson.decode(response, type=ProjectStatus) == ProjectStatus(
-        pid=1, rss_mb=0.1, ready=True, message="ok"
+    await _test_dispatcher_call(
+        dispatcher,
+        {"method": "project-status"},
+        ProjectStatus(pid=1, rss_mb=0.1, ready=True, message="ok"),
```
[the remark this answers is not in the document]
Comment (leynos) [reply weaverd/unittests/test_rpc.py:42]: @coderabbitai What remains to be done for this issue? Please supply an AI agent prompt for the remaining tasks

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

Code Duplication
weaverd/unittests/test_rpc.py
Suppress finding
What lead to degradation?

The module contains 2 functions with similar structure: test_dispatcher_handles_registered_method,test_dispatcher_passes_parameters
Why does this problem occur?

Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.
How to fix it?

A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)

Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

**Comment by leynos** [comment]
@coderabbitai review
